### PR TITLE
Network Costs Prometheus Discovery Scrape Annotations

### PR DESCRIPTION
**Pull request by mbolt35**
https://github.com/kubecost/cost-analyzer-helm-chart/pull/795#issuecomment-799735549 for testing done

### CODE REVIEW

**Comment by dwbrown2** [comment]
cc @michaelmdresser because he's thinking about this in a similarish situation

**Comment by dwbrown2** [comment]
Bolt, were you able to determine if enabling leads to double scraping when our job was already added? 

LGTM, if you are comfortable with how this was tested

File: cost-analyzer/templates/cost-analyzer-network-costs-template.yaml [hunk 1]
```diff
@@ -32,7 +39,7 @@ spec:
 {{- end }}
 {{- if .Values.networkCosts.resources }}
         resources:
-{{ toYaml .Values.networkCosts.resources | indent 10 }}
+          {{ toYaml .Values.networkCosts.resources | indent 10 }}
```
Comment (michaelmdresser) [inline cost-analyzer/templates/cost-analyzer-network-costs-template.yaml:42]: Is this right? There's already an `indent` here.

Comment (mbolt35) [reply cost-analyzer/templates/cost-analyzer-network-costs-template.yaml:42]: That's a good catch, going to blame auto format here, but my tests don't seem to generate something improperly formatted after running `helm template`. 

```yaml
      containers:
      - name: release-name-network-costs
        image: gcr.io/kubecost1/kubecost-network-costs:v15.0
        imagePullPolicy: Always
        resources:
          requests:
            cpu: 50m
            memory: 20Mi
          
        env:
        - name: NODE_NAME
```

I was pretty sure that this was going to be an issue, but both helm2 and helm3 generated the yaml with proper indention. Thoughts?

Comment (mbolt35) [reply cost-analyzer/templates/cost-analyzer-network-costs-template.yaml:42]: Also worth noting that the `indent` command indents _all_ the output from the pipe source. In this case, all of the yaml gets indented by 4, but I would've still expected the first line to be indented further.

Comment (michaelmdresser) [reply cost-analyzer/templates/cost-analyzer-network-costs-template.yaml:42]: Oh, huh. Well, all's well that ends well.

**Comment by michaelmdresser** [comment]
> Bolt, were you able to determine if enabling leads to double scraping when our job was already added?
> 
> LGTM, if you are comfortable with how this was tested

**Comment by dwbrown2** [comment]
@mbolt35 anything left on this PR?

**Comment by pierluigilenoci** [comment]
@mbolt35 when will this PR be released? Because it is not inside the v1.76.0 like you said.

**Comment by mbolt35** [comment]
> @mbolt35 when will this PR be released? Because it is not inside the v1.76.0 like you said.

@pierluigilenoci  Apologies if I mentioned it would make the release. I have some follow up questions here. Based on our helm chart, our prometheus dependency does _not_ have a configuration in place that would consume network costs metrics using these annotations. We would like to add this config, otherwise adding this option to our chart doesn't make sense outside of your specific use case. May I ask what your specific scrape_config looks like to scrape the pods for this annotation?

I drafted the following:
```yaml
- job_name: 'kubernetes-pods'
  kubernetes_sd_configs:
  - role: pod
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape]
    action: keep
    regex: true
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scheme]
    action: replace
    target_label: __scheme__
    regex: (https?)
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_path]
    action: replace
    target_label: __metrics_path__
    regex: (.+)
  - source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
    action: replace
    target_label: __address__
    regex: ([^:]+)(?::\d+)?;(\d+)
    replacement: $1:$2
```

Does your configuration look similar?

**Comment by pierluigilenoci** [comment]
Hi @mbolt35,
my Prometheus configuration is the default one of Prometheus
https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus/values.yaml#L1575-L1607

I don't understand why to ask to add a specific config to Prometheus when the vanilla configuration can be used instead. Forcing users to change the configuration of existing installations instead of adapting the software to work across all configurations doesn't seem like the right approach. 

What I asked you to add is the default annotation that Prometheus uses to read metrics from pods. Why not use it instead of complicating life with configurations that are more difficult to maintain?

**Comment by pierluigilenoci** [comment]
For example, if Kubocost stopped using CAdvisor's deprecated metrics, Prometheus' vanilla configuration would work perfectly for you as well. https://github.com/kubecost/cost-analyzer-helm-chart/issues/747 (Please read the whole discussion carefully.)

**Comment by mbolt35** [comment]
> For example, if Kubocost stopped using CAdvisor's deprecated metrics, Prometheus' vanilla configuration would work perfectly for you as well. #747 (Please read the whole discussion carefully.)

@pierluigilenoci Understood. There are several different approaches that are currently in the mix, and we need to choose the best solution for our out of box install. Unfortunately, it's not as simple given the backwards compatibility requirement we have to support for existing users. It appears that another way this is done is by leveraging the `endpoints` config for node-exporter, which we'd need to move to using a `Service` template for network-costs (annotated with `prometheus.io/scrape`). We're still discussing how this can be implemented. Thanks for your patience!

**Comment by mbolt35** [comment]
> Hi @mbolt35,
> my Prometheus configuration is the default one of Prometheus
> https://github.com/prometheus-community/helm-charts/blob/main/charts/prometheus/values.yaml#L1575-L1607
> 
> I don't understand why to ask to add a specific config to Prometheus when the vanilla configuration can be used instead. Forcing users to change the configuration of existing installations instead of adapting the software to work across all configurations doesn't seem like the right approach.
> 
> What I asked you to add is the default annotation that Prometheus uses to read metrics from pods. Why not use it instead of complicating life with configurations that are more difficult to maintain?
> 
> P.S.: you mentioned here [#770 (comment)](https://github.com/kubecost/cost-analyzer-helm-chart/issues/770#issuecomment-785470253)

Our default prometheus config doesn't have a pod scrape target, which is why I asked. It's also why it's important that we move existing users who use our helm installed prometheus to receive the same update.

**Comment by pierluigilenoci** [comment]
@mbolt35 as I wrote somewhere in another comment I prefer a more general solution lowering the impact of external tool configuration changes. A tool should be as autonomous as possible in its configuration, minimizing the impact on the ecosystem in which it is integrated.

I will quote myself again. 😛 

I can understand that you want to keep backward compatibility but it shouldn't be done at the expense of new versions. In the charts, it is the norm, when some changes break the compatibility to announce the breaking changes and a migration guide. Anyone who wants to upgrade the software but still remains at the legacy version of the cluster can apply the workarounds.

In your chart instead, you impose workarounds and over-configurations on those who keep their clusters updated (or using vanilla configuration) to favor legacy versions. In this way, all those who want to start using the software in updated clusters make a bad experience.

**Comment by mbolt35** [comment]
> @mbolt35 as I wrote somewhere in another comment I prefer a more general solution lowering the impact of external tool configuration changes. A tool should be as autonomous as possible in its configuration, minimizing the impact on the ecosystem in which it is integrated.
> 
> I will quote myself again. 😛
> 
> I can understand that you want to keep backward compatibility but it shouldn't be done at the expense of new versions. In the charts, it is the norm, when some changes break the compatibility to announce the breaking changes and a migration guide. Anyone who wants to upgrade the software but still remains at the legacy version of the cluster can apply the workarounds.
> 
> In your chart instead, you impose workarounds and over-configurations on those who keep their clusters updated (or using vanilla configuration) to favor legacy versions. In this way, all those who want to start using the software in updated clusters make a bad experience.

Thanks for your insight.

**Comment by mbolt35** [comment]
Test:
```
$helm template ./cost-analyzer -f ./cost-analyzer/values.yaml --set networkCosts.enabled=true --set networkCosts.prometheusScrape=true > k.yaml
```

Check Output:
```yaml
$ grep "cost-analyzer-network-costs-service*" k.yaml -A 100
# Source: cost-analyzer/templates/cost-analyzer-network-costs-service.yaml

apiVersion: v1
kind: Service
metadata:
  name: release-name-network-costs
  annotations:
    prometheus.io/scrape: "true"
    prometheus.io/port: "3001"
  labels:
    app: release-name-network-costs
    app.kubernetes.io/name: cost-analyzer
    helm.sh/chart: cost-analyzer-1.76.0
    app.kubernetes.io/instance: release-name
    app.kubernetes.io/managed-by: Tiller
    app: cost-analyzer
spec:
  clusterIP: None
  ports:
  - name: metrics
    port: 3001
    protocol: TCP
    targetPort: 3001
  selector:
    app: release-name-network-costs
  type: ClusterIP
---
```

This configuration will be picked up by our existing default prometheus configuration: 
```
- job_name: 'kubernetes-service-endpoints'
```

**Comment by dwbrown2** [comment]
@mbolt35 I added a corresponding docs change here: https://github.com/kubecost/docs/pull/73

**Review by AjayTripathy — APPROVED** [review]

**Comment by pierluigilenoci** [comment]
❤️

**Comment by anthosz** [comment]
> 



> Test:
> 
> ```
> $helm template ./cost-analyzer -f ./cost-analyzer/values.yaml --set networkCosts.enabled=true --set networkCosts.prometheusScrape=true > k.yaml
> ```
> 
> Check Output:
> 
> ```yaml
> $ grep "cost-analyzer-network-costs-service*" k.yaml -A 100
> # Source: cost-analyzer/templates/cost-analyzer-network-costs-service.yaml
> 
> apiVersion: v1
> kind: Service
> metadata:
>   name: release-name-network-costs
>   annotations:
>     prometheus.io/scrape: "true"
>     prometheus.io/port: "3001"
>   labels:
>     app: release-name-network-costs
>     app.kubernetes.io/name: cost-analyzer
>     helm.sh/chart: cost-analyzer-1.76.0
>     app.kubernetes.io/instance: release-name
>     app.kubernetes.io/managed-by: Tiller
>     app: cost-analyzer
> spec:
>   clusterIP: None
>   ports:
>   - name: metrics
>     port: 3001
>     protocol: TCP
>     targetPort: 3001
>   selector:
>     app: release-name-network-costs
>   type: ClusterIP
> ---
> ```
> 
> This configuration will be picked up by our existing default prometheus configuration:
> 
> ```
> - job_name: 'kubernetes-service-endpoints'
> ```

Hello @mbolt35 ,

It's not really clear how it works.

On my side,  I added networkCosts.prometheusScrape=true + serviceMonitor.networkCosts.enabled=true (same for serviceMonitor.enabled + networkCosts.enabled (services selected)) but what do I need to add on the existing prometheus server to be able to see a cost? (If I replace prometheus by the one included in the kubecost helm chart, it works).

According to the doc, it supposed to be works and no other scrape are mentioned excluding kubecost one...

**Comment by anthosz** [comment]
(fixed)

**Pull request merged**